### PR TITLE
Add user guide documentation for contribution and verification

### PR DESCRIPTION
**Pull request by ping-ke**
---
name: Add user guide
about: Add user guide documentation for contribution and verification
---

# Description

Add user guide documentation:
1. contribution: trusted setup background, requirement, and how to contribute
2. verification: Overview and step-by-step verify

# How Has This Been Tested?
Contribute to the ceremony by following the `contribution.md` documentation, and complete it successfully.
Verify your contribution by following the `verification.md` documentation, and complete it successfully.

### CODE REVIEW

File: docs/verification.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+- Verifying the transcript matches the circuit;
+- Verifying contributors' attestations match the transcript;
+
+## Verifying the Circuit
+Confirming the integrity and security of the [EthStorage blob poseidon circuit](https://github.com/ethstorage/zk-decoder/blob/main/circom/circuits/blob_poseidon_2.circom) 
+by reviewing our circom code, which is open-source and available on Github.
+
+## Verifying the Transcript Matches the Circuit
+Our ceremony transcript is included in the zkey published on ceremony.ethstorage.io in the "Download ZKey" tab of the 
+[EthStorage Circom page](https://ceremony.ethstorage.io/projects/EthStorage%20circom). You can verify it matches the
```
Comment (qzhodl) [inline docs/verification.md:27]: Still an invalid url

File: docs/contribution.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+```bash
+npm install @ethstorage-org/phase2cli
+```
+
+3. **Authenticate with GitHub**
+
+If your GitHub meets the requirements, you will be eligible for this next step using the following command
+```bash
+phase2cli auth
```
Comment (qzhodl) [inline docs/contribution.md:94]: error: command not found: phase2cli

File: docs/verification.md (hunk 1 above)
Comment (ping-ke) [reply docs/verification.md:27]: it should be ok now.

File: docs/contribution.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+the ceremony or a predetermined end date is reached. Ultimately, the ceremony results in the extraction of a key 
+for the specific circuit being used.
+
+
+## Requirements
+
+1. **A computer**: 
+
+A computer with 8GB or above Space and Ram is recommended. Ideally Mac or Linux, Windows via 
+[WSL](https://learn.microsoft.com/en-us/windows/wsl/install) is Ok;
```
Comment (syntrust) [inline docs/contribution.md]: why end with `;`?

File: docs/contribution.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+2. **A strong internet connection**:
+
+The most common cause of failures is an inability to upload your contribution to the ceremony before the timeout 
+period expires. Please run the ceremony with a strong internet connection, and your upload and download bandwidths 
+should be at least 25 Mbps each (preferably 50+ Mbps each). Using a wired (ethernet) connection rather than a 
+wireless (WiFi) connection is also recommended.
+
+3. **A GitHub Accout**:
```
Comment (syntrust) [inline docs/contribution.md]: Accout => Account

File: docs/contribution.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+**Error: Your contribution took longer than the estimated time**: 
+
+If you meet the following error: 
+"Your contribution took longer than the estimated time and you were removed as current contributor. You should wait for 
+    a timeout to expire before you can rejoin for contribution." This means that your contribution uses twice times 
+    contributions average used. Recommend using better machines (hardware and network) to provide contribution.
+
+**If you disconnect**: While in the queue, you can re-join by running the npx phase2cli contribute command again.
+
+**Authentication errors**: phase2cli authentication can get rejected if you initially failed the GitHub requirements.
```
Comment (syntrust) [inline docs/contribution.md]: GitHub=>GitHub account

File: docs/contribution.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+![clean.png](clean.png)
+![logout.png](logout.png)
+
+## Note
+**Make sure your computer does not go to sleep**: When your turn in the queue comes, your computer needs to be awake 
+    and connected to the internet. If you’re going to leave your contribution running and step away from the computer, 
+    be sure you don’t have your computer set to automatically sleep.
+
+
+## FAQ
```
Comment (syntrust) [inline docs/contribution.md]: Better title "Frequently Encountered Errors" than FAQ?

File: docs/contribution.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+<img alt="auth1.png" src="auth1.png" width=40%/>
+
+4. **Contribute**
+
+```bash
+phase2cli contribute
+```
+
+Run the command and select the circom you want to contribute (like _EthStorage circom_);
```
Comment (syntrust) [inline docs/contribution.md]: Are there multiple choices in a product environment as in the screenshot? Need explanations for each if yes.

File: docs/verification.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+Confirming the integrity and security of the [EthStorage blob poseidon circuit](https://github.com/ethstorage/zk-decoder/blob/main/circom/circuits/blob_poseidon_2.circom) 
+by reviewing our circom code, which is open-source and available on Github.
+
+## Verifying the Transcript Matches the Circuit
+Our ceremony transcript is included in the zkey published on ceremony.ethstorage.io in the "Download ZKey" tab of the 
+[EthStorage Circom page](https://ceremony.ethstorage.io/projects/EthStorage%20circom). You can verify it matches the 
+circuit using Circom and snarkjs:
+
+- Install [Circom](https://docs.circom.io/getting-started/installation/) and [snarkjs](https://github.com/iden3/snarkjs).
+- Clone circom code https://github.com/ethstorage/zk-decoder.
```
Comment (syntrust) [inline docs/verification.md:31]: `https://github.com/ethstorage/zk-decoder` is a private repo

File: docs/verification.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+that malicious projects pretend to run rituals while secretly retaining backdoors. 
+To ensure that anyone can verify the integrity of our trusted setup ceremony, we have outlined a robust guide for 
+verifying the results. Here’s an overview of the key steps involved:
+
+- Verifying the circuit we are securing does not include security holes;
+- Verifying the transcript matches the circuit;
+- Verifying contributors' attestations match the transcript;
+
+## Verifying the Circuit
+Confirming the integrity and security of the [EthStorage blob poseidon circuit](https://github.com/ethstorage/zk-decoder/blob/main/circom/circuits/blob_poseidon_2.circom)
```
Comment (syntrust) [inline docs/verification.md:22]: `https://github.com/ethstorage/zk-decoder` is a private repo, not sure everyone can see the code

File: docs/verification.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+## Verifying the Transcript Matches the Circuit
+Our ceremony transcript is included in the zkey published on ceremony.ethstorage.io in the "Download ZKey" tab of the 
+[EthStorage Circom page](https://ceremony.ethstorage.io/projects/EthStorage%20circom). You can verify it matches the 
+circuit using Circom and snarkjs:
+
+- Install [Circom](https://docs.circom.io/getting-started/installation/) and [snarkjs](https://github.com/iden3/snarkjs).
+- Clone circom code https://github.com/ethstorage/zk-decoder.
+- Download the Powers of Tau (ptau) file we used for our ceremony; we used the Hermez rollup with 2^20 powers of tau, 
+which is linked in [the snarkjs readme](https://github.com/iden3/snarkjs) or available directly 
+[here](https://storage.googleapis.com/zkevm/ptau/powersOfTau28_hez_final_20.ptau).
```
Comment (qzhodl) [inline docs/verification.md]: Need to update the new link here?

File: docs/verification.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+This proves which GitHub user made each contribution to the ceremony. You can verify attestations by looking them 
+up both in the transcript and on GitHub:
+
+In the transcript (generated by snarkjs in the previous step), you will see that each contribution is preceded 
+by the phrase contribution `#[number] [username]-[number]`. That username is the GitHub username.
+
+On GitHub, you can run a search to find an attestation Gist on gist.github.com. For example, if the username is 
+`somecontributor`, then you could search for filename:`"ethstorage-circom_attestation.log" user:somecontributor`.
+
+To verify an attestation, confirm that the hash in the attestation in the transcript matches the hash in the
```
Comment (qzhodl) [inline docs/verification.md:59]: Can not find Ping-Ke's first contribution on the gist

Comment (ping-ke) [reply docs/verification.md:59]: That is because it is create by ceremony setup

File: docs/contribution.md (hunk 7 above)
Comment (ping-ke) [reply docs/contribution.md]: Circoms already have explanations when users are selecting.

File: docs/verification.md (hunk 11 above)
Comment (qzhodl) [reply docs/verification.md:59]: But it should be available on the GitHub Gist, right?

File: docs/verification.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+Our ceremony transcript is included in the zkey published on ceremony.ethstorage.io in the "Download ZKey" tab of the 
+[EthStorage Circom page](https://ceremony.ethstorage.io/projects/EthStorage%20circom). You can verify it matches the 
+circuit using Circom and snarkjs:
+
+- Install [Circom](https://docs.circom.io/getting-started/installation/) and [snarkjs](https://github.com/iden3/snarkjs).
+- Clone circom code https://github.com/ethstorage/zk-decoder.
+- Download the Powers of Tau (ptau) file we used for our ceremony; we used the Hermez rollup with 2^20 powers of tau, 
+which is linked in [the snarkjs readme](https://github.com/privacy-scaling-explorations/perpetualpowersoftau?tab=readme-ov-file#prepared-phase-2-files) 
+or available directly [here](https://pse-trusted-setup-ppot.s3.eu-central-1.amazonaws.com/pot28_0080/ppot_0080_20.ptau).
+- Generate the `r1cs` file from the `circom` source files, by running `circom blob_poseidon_2.circom --r1cs` in the
```
Comment (syntrust) [inline docs/verification.md:35]: Missed the step to install dependencies?
```
error[P1014]:  The file ../node_modules/circomlib/circuits/poseidon.circom to be included has not been found
 = Consider using compilation option -l to indicate include paths
```

**Review by qzhodl — APPROVED** [review]

File: docs/verification.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+[EthStorage Circom page](https://ceremony.ethstorage.io/projects/EthStorage%20circom). You can verify it matches the 
+circuit using Circom and snarkjs:
+
+- Install [Circom](https://docs.circom.io/getting-started/installation/) and [snarkjs](https://github.com/iden3/snarkjs).
+- Clone circom code https://github.com/ethstorage/zk-decoder.
+- Download the Powers of Tau (ptau) file we used for our ceremony; we used the Hermez rollup with 2^20 powers of tau, 
+which is linked in [the snarkjs readme](https://github.com/privacy-scaling-explorations/perpetualpowersoftau?tab=readme-ov-file#prepared-phase-2-files) 
+or available directly [here](https://pse-trusted-setup-ppot.s3.eu-central-1.amazonaws.com/pot28_0080/ppot_0080_20.ptau).
+- Generate the `r1cs` file from the `circom` source files, by running `circom blob_poseidon_2.circom --r1cs` in the 
+directory `zk-decoder/circom/circuits` (need to run `npm install` for the first time). The SHA-256 hash of this file
```
Comment (syntrust) [inline docs/verification.md:36]: It's confusing about where to run `npm install`

File: docs/verification.md (hunk 1 above)
Comment (syntrust) [inline docs/verification.md:27]: It would be more convenient for cmd users to provide a URL to download zkey.

File: docs/verification.md (hunk 13 above)
Comment (ping-ke) [reply docs/verification.md:36]: under zk-decoder/circom/circuits, same folder of `circom blob_poseidon_2.circom --r1cs`

Comment (syntrust) [reply docs/verification.md:36]: Actually `npm install` should be run under `zk-decoder/circom` and it should be told clearly

File: docs/verification.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+(i.e. as computed by `sha256sum blob_poseidon_2.r1cs`) is `ba17a94a5edcc1d0b0ad10990e0aea4894f080ce571b0db79fb1c3a9c4a74a0d`
+- Use snarkjs to verify that the transcript matches this circuit and powers of tau, by running 
+`snarkjs zkey verify blob_poseidon_2.r1cs ppot_0080_20.ptau blob_poseidon_2_final.zkey`. `blob_poseidon_2_final.zkey` can be downloaded
```
Comment (syntrust) [inline docs/verification.md:39]:
```
# snarkjs zkey verify blob_poseidon_2.r1cs ppot_0080_20.ptau blob_poseidon_2_final.zkey
[INFO]  snarkJS: Reading r1cs
[INFO]  snarkJS: Reading tauG1
[INFO]  snarkJS: Reading tauG2
[INFO]  snarkJS: Reading alphatauG1
[INFO]  snarkJS: Reading betatauG1

[INFO]  snarkJS: Circuit hash: 
                15c4f4f6 1804b492 ca420a19 8e94ffb5
                6f90f0bd 5241de8a 417534fb 13f2a6cb
                11e69823 8968e462 f50fc24b f15fdf9e
                e78471d8 7599abd3 5b12a428 33e1f184
[ERROR] snarkJS: DOMException [QuotaExceededError]: The requested length exceeds 65,536 bytes
    at new DOMException (node:internal/per_context/domexception:53:5)
    at __node_internal_ (node:internal/util:695:10)
    at Crypto.getRandomValues (node:internal/crypto/random:328:11)
    at Crypto.getRandomValues (node:internal/crypto/webcrypto:999:10)
    at getRandomBytes (/root/.nvm/versions/node/v20.10.0/lib/node_modules/snarkjs/build/cli.cjs:285:27)
    at sectionHasSameRatio (/root/.nvm/versions/node/v20.10.0/lib/node_modules/snarkjs/build/cli.cjs:4768:29)
    at async phase2verifyFromInit (/root/.nvm/versions/node/v20.10.0/lib/node_modules/snarkjs/build/cli.cjs:4716:10)
    at async phase2verifyFromR1cs (/root/.nvm/versions/node/v20.10.0/lib/node_modules/snarkjs/build/cli.cjs:4960:12)
    at async Object.zkeyVerifyFromR1cs [as action] (/root/.nvm/versions/node/v20.10.0/lib/node_modules/snarkjs/build/cli.cjs:13442:17)
    at async clProcessor (/root/.nvm/versions/node/v20.10.0/lib/node_modules/snarkjs/build/cli.cjs:481:27)

```

File: docs/verification.md (hunk 13 above)
Comment (ping-ke) [reply docs/verification.md:36]: both `zk-decoder/circom` and `zk-decoder/circom/circuits` are ok

Comment (syntrust) [reply docs/verification.md:36]: ok

**Review by syntrust — APPROVED** [review]

**Pull request merged**